lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -166,28 +201,91 @@ def _on_peer_relation_joined(self, event: RelationJoinedEvent):
 
     def _on_peer_relation_changed(self, event: RelationChangedEvent):
         """Handle peer relation changes."""
+        data = event.relation.data.get(event.unit if self.unit.is_leader() else event.app)
+        if not data:
+            return
+
         if self.unit.is_leader():
-            data = event.relation.data.get(event.unit)
-            if data:
-                exclusions_to_remove = data.get("remove_from_allocation_exclusions")
-                if exclusions_to_remove:
-                    self.append_allocation_exclusion_to_remove(exclusions_to_remove)
-
-                if data.get("bootstrap_contributor"):
-                    contributor_count = self.peers_data.get(
-                        Scope.APP, "bootstrap_contributors_count", 0
-                    )
-                    self.peers_data.put(
-                        Scope.APP, "bootstrap_contributors_count", contributor_count + 1
-                    )
+            cluster_health_color = data.get("health", "green")
+            if cluster_health_color != "green":
+                self.app.status = BlockedStatus(
+                    ClusterHealthRed if cluster_health_color == "red" else ClusterHealthYellow
+                )
 
-        # Restart node when cert renewal for the transport layer
-        if self.peers_data.get(Scope.UNIT, "must_reboot_node"):
-            try:
-                self.opensearch.restart()
-                self.peers_data.delete(Scope.UNIT, "must_reboot_node")
-            except OpenSearchStartError:
-                event.defer()
+            if data.get("bootstrap_contributor"):
+                contributor_count = self.peers_data.get(
+                    Scope.APP, "bootstrap_contributors_count", 0
+                )
+                self.peers_data.put(
+                    Scope.APP, "bootstrap_contributors_count", contributor_count + 1
+                )
+
+            if data.get(VOTING_TO_DELETE) or data.get(ALLOCS_TO_DELETE):
+                self.opensearch_exclusions.cleanup()
+
+        # Run restart node on the concerned unit
+        self._restart_unit_with_conf_if_needed()
+
+    def _on_peer_relation_departed(self, event: RelationDepartedEvent):
```

Comment:
I'm in favor of it, it's getting quite big. And it would also be helpful to work with jsonschema to describe everything we store in the peer relation. I'll create a ticket for that